tools/lkl/tests/hijack-test.sh

```diff
@@ -1,591 +1,741 @@
+                "masklen":"$TEST_IP_NETMASK",
+                "mac":"$TEST_MAC0",
+                "ipv6":"$(ip6_host)",
+                "masklen6":"$TEST_IP6_NETMASK"
+            }
+        ]
+    }
 EOF
 
-# Need to say || true because ip -h returns < 0
-ans=$(LKL_HIJACK_CONFIG_FILE=$cfgjson\
-  ${hijack_script} ip -h) || true
-# Need to grab the end because something earlier on prints out this
-# number
-# XXX: android-23 doesn't call destructor...
-if [ -z ${LKL_ANDROID_TEST} ] ; then
-echo "$ans" | tail -n 15 | grep "65536" # lo's MTU
-# lo's dev id
-echo "$ans" | grep "0x0"        # lo's dev_id
-# Doesn't really belong in this section, but might as well check for
-# it here.
-! echo "$ans" | grep "WARN: failed to free"
-fi
+    ${hijack_cfg} $(lkl_test_cmd which sleep) 10 &
+
+    set_cfgjson << EOF
+    {
+        "gateway":"$(ip_host)",
+        "gateway6":"$(ip6_host)",
+        "interfaces":
+        [
+            {
+                "type":"pipe",
+                "param":"${fifo2}|${fifo1}",
+                "ip":"$(ip_lkl)",
+                "masklen":"$TEST_IP_NETMASK",
+                "mac":"$TEST_MAC0",
+                "ipv6":"$(ip6_lkl)",
+               "masklen6":"$TEST_IP6_NETMASK"
+            }
+        ]
+    }
+EOF
 
-# boot_cmdline test
-echo "== boot command line tests =="
+    # Ping under LKL
+    ${hijack_cfg} ${ping} $(ip_host) -c 1 -w 10
 
-cat << EOF > ${cfgjson}
-{
-	"debug":"1",
-	"boot_cmdline":"mem=100M"
+    # Ping 6 under LKL
+    ${hijack_cfg} ${ping6} $(ip6_host) -c 1 -w 10
+
+    jobs
+    kill -9 %1
+    wait
 }
+
+test_tap_setup()
+{
+    set -e
+
+    # Set up the TAP device we'd like to use
+    tap_setup
+
+    set_cfgjson << EOF
+    {
+        "gateway":"$(ip_host)",
+        "gateway6":"$(ip6_host)",
+        "debug":"1",
+        "interfaces": [
+            {
+                "type":"tap",
+                "param":"$(tap_ifname)",
+                "ip":"$(ip_lkl)",
+                "masklen":"$TEST_IP_NETMASK",
+                "ipv6":"$(ip6_lkl)",
+                "masklen6":"$TEST_IP6_NETMASK",
+                "mac": "$TEST_MAC0"
+            }
+        ]
+    }
 EOF
 
-ans=$(LKL_HIJACK_CONFIG_FILE=$cfgjson ${hijack_script} ip ad)
-if [ -z ${CROSS_COMPILE} ] ; then
-echo "$ans" | grep "100752k"
-elif [ "${CROSS_COMPILE}" = "arm-linux-androideabi-" ] ; then
-(echo "$ans" | grep "101424k") || true
-elif [ "${CROSS_COMPILE}" = "aarch64-linux-android-" ] ; then
-(echo "$ans" | grep "100756k") || true
-fi
+    # Make sure our device has the addresses we expect
+    addr=$(${hijack_cfg} ip addr)
+    echo "$addr" | grep eth0
+    echo "$addr" | grep $(ip_lkl)
+    echo "$addr" | grep "$TEST_MAC0"
+    echo "$addr" | grep "$(ip6_lkl)"
+    ! echo "$addr" | grep "WARN: failed to free"
+}
 
-echo "== PIPE tests =="
-if [ -z `which mkfifo` ]; then
-    echo "WARNIG: no mkfifo command, skipping PIPE tests."
-else
+test_tap_cleanup()
+{
+    tap_cleanup
+    tap_cleanup 1
+}
+
+test_tap_ping_host()
```

Comment:
I don't know how it works on android.
IIRC android doesn't have tap support but the following test results says `TEST       ok       test_tap_ping_host `.  

https://circleci.com/gh/tavip/linux/134?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-build-link

Reply:
It does work, it just needs the right permissions, which in this case are: user=vpn,group=vpn and additional group=net_admin. See the tap_prepare and tap_setup in net-setup.sh

Reply:
that's new to me (cc: @phhusson @linhua55 and #379). Great and thanks !